### PR TITLE
TST: Add tests for fixed issues

### LINKED ISSUE #5286: Phantom Index when NaN in MultiIndex
Using the following:

```
csv_str = """u1,u2,u3,d1,d2,d3,d4
Good Things,C,,1,1,1,1
Good Things,R,,1,1,1,1
Bad Things,C,,1,1,1,1
Bad Things,T,,1,1,1,1
Okay Things,N,B,1,1,1,1
Okay Things,N,D,1,1,1,1
Okay Things,B,,1,1,1,1
Okay Things,D,,1,1,1,1"""
d_nan = pd.read_csv(StringIO.StringIO(csv_str), na_filter=False,index_col=[0,1,2])
d_no_nan = pd.read_csv(StringIO.StringIO(csv_str), na_filter=False,index_col=[0,1,2],na_values=[], keep_default_na=False)
```

If I have the following DataFrame:

```
In [63]: d_nan
Out[63]: 
                    d1  d2  d3  d4
u1          u2 u3                 
Good Things C  NaN   1   1   1   1
            R  NaN   1   1   1   1
Bad Things  C  NaN   1   1   1   1
            T  NaN   1   1   1   1
Okay Things N  B     1   1   1   1
               D     1   1   1   1
            B  NaN   1   1   1   1
            D  NaN   1   1   1   1
```

and then I try to use `loc` on the index:

```
In [64]: d_nan.loc["Good Things"]
Out[64]: 
        d1  d2  d3  d4
u2 u3                 
C  NaN   1   1   1   1
R  NaN   1   1   1   1
```

When I try to `loc` the last level, a mysterious D appears:

```
In [65]: d_nan.loc["Good Things"].loc['C']
Out[65]: 
    d1  d2  d3  d4
u3                
D    1   1   1   1
```

if d instead looks like (no NaNs):

```
In [67]: d_no_nan
Out[67]: 
                   d1  d2  d3  d4
u1          u2 u3                
Good Things C       1   1   1   1
            R       1   1   1   1
Bad Things  C       1   1   1   1
            T       1   1   1   1
Okay Things N  B    1   1   1   1
               D    1   1   1   1
            B       1   1   1   1
            D       1   1   1   1
```

then I have no problems:

```
In [68]: d_no_nan.loc["Good Things"].loc['C']
Out[68]: 
    d1  d2  d3  d4
u3                
     1   1   1   1
```

### LINKED ISSUE #5586: resample with numpy.product gives unexpected results
```
import datetime
import numpy
import pandas

index = pandas.DatetimeIndex(start=datetime.date(2012, 1, 31), freq='M', periods=12)

ts = pandas.Series(range(12), index=index)
df = pandas.DataFrame(dict(A=ts, B=ts+2))

print df

# Correct
print df.resample('M', how=numpy.sum)
# Incorrect
print df.resample('M', how=numpy.product)
# Correct
print df.resample('Q', how=numpy.sum)
# Incorrect
print df.resample('Q', how=numpy.product)
```

for print df.resample('Q', how=numpy.product), was expecting:

```
2012-03-31   2*1*0   4*3*2
2012-06-30   5*4*3   7*6*5
2012-09-30   8*7*6  10*9*8
2012-12-31  11*10*9  13*12*11
```

### LINKED ISSUE #5647: 0.12: join with on= fails with TimeStamp and Int64 MultiIndex
I encountered a very strange problem in 0.12 when a straightforward join fails while it works just fine in 0.11.

```
df1
```

> <class 'pandas.core.frame.DataFrame'>
> Int64Index: 71958 entries, 0 to 104014
> Data columns (total 3 columns):
> FactorDate         71958  non-null values
> FundSecNo          71958  non-null values
> SourceFundSecNo    71958  non-null values
> dtypes: datetime64[ns](1), int64(2)

```
df1.dtypes
```

> FactorDate         datetime64[ns]
> FundSecNo                   int64
> SourceFundSecNo             int64
> dtype: object

```
df2
```

> <class 'pandas.core.frame.DataFrame'>
> MultiIndex: 94868 entries, (2011-10-18 00:00:00, 283) to (2011-10-19 00:00:00, 127419)
> Data columns (total 2 columns):
> DemeanGroupId           94868  non-null values
> NormalizationGroupId    94868  non-null values
> dtypes: float64(1), int64(1)

```
for ind in df2.index.levels:
    print(type(ind))
```

> <class 'pandas.tseries.index.DatetimeIndex'>
> <class 'pandas.core.index.Int64Index'>

```
test = df1.join(df2, on=['FactorDate', 'SourceFundSecNo'])
test
```

> <class 'pandas.core.frame.DataFrame'>
> Int64Index: 71958 entries, 0 to 104014
> Data columns (total 5 columns):
> FactorDate              71958  non-null values
> FundSecNo               71958  non-null values
> SourceFundSecNo         71958  non-null values
> DemeanGroupId           0  non-null values
> NormalizationGroupId    0  non-null values
> dtypes: datetime64[ns](1), float64(2), int64(2)

```
date = datetime.datetime(2011, 10, 18)
sfsn = 283
test.loc[(test.FactorDate == date) & (test.SourceFundSecNo == sfsn)]
```

> ```
>           FactorDate  FundSecNo  SourceFundSecNo  DemeanGroupId  NormalizationGroupId
> ```
> 
> 1     2011-10-18 00:00:00        283              283            NaN                   NaN
> 17933 2011-10-18 00:00:00      65213              283            NaN                   NaN

```
df2.loc[(date, sfsn)]
```

> DemeanGroupId           350171
> NormalizationGroupId        35
> Name: (2011-10-18 00:00:00, 283), dtype: float64

I can attach the frames, but I forgot how to do it :-)

### LINKED ISSUE #6485: BUG/ENH: dtype change on enlargement
http://stackoverflow.com/questions/22044766/adding-row-to-pandas-dataframe-changes-dtype/22045011#22045011

A bit non-trival (maybe coerce back after) if possible?

maybe just change this to an actual append?

```
In [1]: df = DataFrame({'a' : range(10)}, dtype='i4')

In [2]: df
Out[2]: 
   a
0  0
1  1
2  2
3  3
4  4
5  5
6  6
7  7
8  8
9  9

[10 rows x 1 columns]

In [3]: df.dtypes
Out[3]: 
a    int32
dtype: object

In [4]: df.loc[10] = 99

In [5]: df
Out[5]: 
     a
0    0
1    1
2    2
3    3
4    4
5    5
6    6
7    7
8    8
9    9
10  99

[11 rows x 1 columns]

In [6]: df.dtypes
Out[6]: 
a    float64
dtype: object

```

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #5647
- [x] closes #5586
- [x] closes #5286
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by jbrockmendel** [comment]
looks like a int32 vs int64 issue?

**Comment by mroeschke** [comment]
@jbrockmendel looks like that was the issue. All green

File: pandas/tests/indexing/test_loc.py [hunk 1]
```diff
@@ -968,3 +968,11 @@ def test_loc_getitem_label_list_integer_labels(
     expected = df.iloc[:, expected_columns]
     result = df.loc[["A", "B", "C"], column_key]
     tm.assert_frame_equal(result, expected, check_column_type=check_column_type)
+
+
+def test_loc_upcast_int():
+    # GH 6485
+    result = DataFrame({"a": range(10)}, dtype="i4")
+    result.loc[10] = 99
+    expected = DataFrame({"a": list(range(10)) + [99]})
```
Comment (jbrockmendel) [inline pandas/tests/indexing/test_loc.py]: can you specify the expected dtype here?  im actually not clear on why this is "fixed".  casting to int64 is better than casting to float64 like it did in the OP in #6485, but dont we want the dtype to stay int32?

Also, there is mention in that issue of an empty case that is still broken.  should that issue stay open or new issue for the still-broken case?

**Comment by jbrockmendel** [comment]
the other 3 LGTM

File: pandas/tests/indexing/test_loc.py (hunk 1 above)
Comment (mroeschke) [reply pandas/tests/indexing/test_loc.py]: Good point. Ideally we should maintain int32 here. I guess that issue is not totally solved and it leave it open for both cases

**Comment by jreback** [comment]
thanks @mroeschke

**Pull request merged**